### PR TITLE
[kubernetes] implement leader election among agent for event collection

### PR DESCRIPTION
**Pull request by hkaj**
*Note: Please remember to review the Datadog [Contribution Guidelines](https://github.com/DataDog/dd-agent/blob/master/CONTRIBUTING.md)
if you have not yet done so.*

### What does this PR do?

Implement a leader election mechanism among agents on kubernetes to collect events automatically.

### Motivation

It's a pain to create a single deployment for the one agent responsible to collect events. It's also not recommended to have all agents report events due to the load this puts on the apiserver (and the redundant data sent by all agents).

This feature will allow having event collection enabled only once, without needed a snowflake agent deployment. All agents can be started with the `leader_candidate: true` option and take part in this election. Only the leading agent will report events.

### Testing Guidelines

Unit tests were added, integration tests are err... manual for now.

### Additional Notes

Linked with https://github.com/DataDog/integrations-core/pull/687

### CODE REVIEW

File: utils/kubernetes/leader_elector.py [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+            if ex.response.status_code == 404:
+                return None
+            log.error("Failed to get config map %s. Error: %s" % (CM_NAME, str(ex)))
+            return
+        if len(res) == 0:
+            return None
+        elif len(res) == 1:
+            cm = res[0]
+            acquired_time = cm['metadata'].get('labels', {}).get(ACQUIRE_TIME_LABEL)
+            self.last_acquire_time = datetime.datetime.strptime(acquired_time, "%Y-%m-%dT%H:%M:%S.%f")
```
Comment (xvello) [inline utils/kubernetes/leader_elector.py]: Do we needed it to be human-readable or could we use a timestamp instead?

File: utils/kubernetes/kube_event_retriever.py [hunk 2]
```diff
@@ -69,8 +69,11 @@ def set_kinds(self, kinds):
     def get_event_array(self):
         """
         Fetch latest events from the apiserver for the namespaces and kinds set on init
-        and returns an array of event objects
+        and returns an array of event objects.
+        Also triggers leader election.
         """
+        # Leader election
+        self.kubeutil.refresh_leader()
```
Comment (hkaj) [inline utils/kubernetes/kube_event_retriever.py]: You're right

File: utils/kubernetes/kubeutil.py [hunk 3]
```diff
@@ -108,6 +108,13 @@ def __init__(self, instance=None):
 
         self.kubernetes_api_url = '%s/api/v1' % self.kubernetes_api_root_url
 
+        # leader status triggers event collection
+        self.is_leader = False
+        self.leader_elector = None
+        if os.environ.get('DD_LEADER_CANDIDATE'):
```
Comment (hkaj) [inline utils/kubernetes/kubeutil.py]: yeah will add something in datadog.conf, this is temporary

File: tests/core/test_leader_elector.py [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+
+# 3rd party
+import mock
+
+# project
+from utils.kubernetes import LeaderElector
+from tests.core.test_kubeutil import KubeTestCase
+
+
+HEALTH_ENDPOINT = '/healthz'
```
Comment (xvello) [inline tests/core/test_leader_elector.py]: Shoud you import the contants from the `LeaderElector` class instead of copying them?

**Comment by xvello** [comment]
Several tests fail because we mocked `retrieve_json_auth`'s response with a dict. We could reuse [this MockResponse object](https://github.com/DataDog/dd-agent/blob/master/tests/core/test_orchestrator.py#L14)

**Pull request merged**